doc/man/radwho.1
.\" $Id$
.ds RD /usr/local/etc/raddb
.ds RL /usr/adm
.ds RP /usr/local/etc/raddb
.TH radwho 1 "March 10, 2000" "FSF"
.SH NAME
radwho - display who is logged on by the Radius

.SH SYNOPSIS
\fBradwho\fR
[\fB-A\fR, \fB--all\fR]
[\fB-c\fR, \fB--calling-id\fR]
[\fB-d,--directory\fR DIR]
[\fB-D,--date-format\fR {short|abbr|full}]
[\fB-e,--empty\fR STRING]
[\fB-f,--file\fR FILE]
[\fB-F,--finger\fR]
[\fB-h,--help\fR]
[\fB-H,--no-header\fR]
[\fB-i,--session-id\fR]
[\fB-I,--ip-format\fR {smart|ip|nodomain}]
[\fB-o,--format\fR FORMAT]
[\fB-s,--secure\fR]


.SH DESCRIPTION
\fBRadwho\fR displays the list of users currently logged in by the
\fBRadius\fR server.
.PP
Default output information is made compatible
with that of the standard unix finger(1) utility. For each user the
following information is displayed: login name, name, connection
protocol, NAS port, login date, NAS name, assigned IP address or
corresponding network name.
.PP
When used with \fB-l\fR option, the long output format is used. In
this format the following information is output:
.IP Login
Login name of the user
.IP SessionID
Unique session ID assigned by the terminal server.
.IP Proto
Connection prototype.
.IP Port
Port number
.IP When
Login date and time
.IP From
The name of NAS that accepted the connection.
.IP Location
Framed IP address or the corresponding network name.
.IP Caller
Caller station ID ad reported by the NAS.
.IP Duration
Duration of the session.

.SH OPTIONS
.IP "-A, --all"
Display the information about logged-out users as well. The logged-out
users are shown with Proto field set to HUP.

.IP "-c, --calling-id"
Display the calling station ID in the second column.

.IP "-d, --directory NAME"
Set the radius configuration directory name.

.IP "-D, --date-formap {short|abbr|full}"
Set the date representation. By default dates are output as
\fBDOW HH:MM\fR, where DOW means the day of week abbreviation,
HH and MM mean hours and minutes respectively. This corresponds to
option \fB-D short\fR. Other available formats are: \fBabbr\fR for
abbreviated date: \fBMM/DD HH:MM\fR, where MM is a two-digit month number,
DD --- a two-digit day of month, and \fBfull\f for full data output,
like this: \fBMon Dec 18 12:29:38 EET 2000\fR.

.IP "-e, --empty STRING"
Display any empty field as STRING. This is useful when the output of
\fBradwho\fR is fed to some analyzing program, as it helps to keep
the same number of columns on each line of output.

.IP "-F, --finger"
Start in fingerd mode. In this mode \fBradwho\fR emulates the
behaviour of fingerd(8) utility. Use this option if starting
\fBradwho\fR from the /etc/inetd.conf line like this:
.IP
finger  stream  tcp   nowait  nobody   /usr/sbin/radwho radwho -fL
.IP
This mode is also enabled by default if \fBradwho\fR notices that its
name (argv[0]) is \fBfingerd\fR or \fBin.fingerd\fR.

.IP "-H, --no-header"
Don't display header line.

.IP "-i, --session-id"
Display session ID instead of GECOS in the second column.

.IP "-I, --ip-format" {smart|ip|nodomain}
Change IP address representation. 

.IP "-L, --local-also"
Display information about local users from the system \fButmp\fR file.
May prove useful when running \fBradwho\fR as a finger daemon.

.IP "-n, --no-resolve"
Do not resolve IP addresses. It is a synonim for \fB-I ip\fR.

.IP "-o, --format FORMAT"
Select customized output format. This can also be changed by setting
the value of environment variable \fBRADWHO_FORMAT\fR. See \fBFORMAT STRING\fR
section below.

.IP "-s, --secure"
Run in secure mode. Queries without a user name are rejected.

.SH FORMAT STRING

.PP
The format string is a comma-separated list of format specifications
in one of the following forms:

.IP \fBfield\fR
Output \fBfield\fR with its default width, heading and alignment. The
field names and corresponding default values are discussed in detail below.
.IP \fBfield\fR:\fBwidth\fR
Output \fBfield\fR, use column width \fBwidth\fR. If \fBwidth\fR starts
with a plus sign, the field will be right-aligned, if it starts with
a minus sign, the field will be left-aligned. Otherwise the default
alignment will be used
.IP \fBfield\fR:\fBwidth\fR:\fBheading\fR
The same as above, but also supplies the heading for the column
.PP
The field names are:

.IP login
Login name

.IP orig
Original login name as supplied with the request.

.IP port
NAS port number

.IP sid
The Session ID

.IP nas
The NAS name or IP address.

.IP ip
Framed IP address assigned to the user, if it is provided framed
service.

.IP proto
Connection protocol. Its possible values are:

.IP date
Date/time when the session started

.IP delay
Delay time.

.IP type
Entry type in decimal.

.IP ptype
Port type. This is one of the following letters:

.IP time
Total time of the session duration.

.IP clid
The calling station ID.

.IP uname
The GECOS field from local /etc/passwd, corresponding to the login name.
If the user does not have a local account, his login name is output.



.SH FILES
.IP \*(RD/naslist
.IP \*(RL/radutmp

.SH BUGS
The information in this manpage may be obsolete or incomplete. Please
refer to texinfo documentation for full information about GNU Radius
tools.

.SH AUTHORS
Sergey Poznyakoff, gray@farlep.net

.SH SEE ALSO
radiusd(8rad), radgrep(1rad),
radping(8rad), naslist(5rad)


NEWS
GNU Radius NEWS -- history of user-visible changes. 2004-04-23
Copyright (C) 2002,2003,2004 Free Software Foundation, Inc.
See the end of file for copying conditions.

Please send radius bug reports to <bug-gnu-radius@gnu.org>
^L
Version 1.2.3:

* libgnuradius

This is a library of functions for creation, handling and sending
requests using RADIUS protocol.

All programs have been rewritten to link with libgnuradius. On most
sites this will mean linking against a shared library, which will
reduce the size of the executables.

* gnuradius.scm

This is a guile module allowing to use libgnuradius functions. It
supercedes radscm program, which has been removed.

* Rewrite language

** Added i18n support

** New built-in functions:

*** Functions to access internal fields of a RADIUS request.
*** Interfaces to the Radius NAS database (raddb/naslist).
*** Interfaces to DNS lookup functions.

* radiusd

** Improved checking for multiple logins. Previous versions relied entirely
on the contents of /var/log/radutmp file. Starting at this version,
radiusd offers at least two methods of checking for multiple logins:
using the traditional radutmp file and using the SQL database. New
keywords has been added to the sqlserver file that declare the SQL
queries to be used when retrieving the information about currently
active sessions.

More methods of checking will be added in future versions.

** When an unsupported authentication type is requested, radiusd first
checks if an extension Scheme module is provided that handles that
authentication type. If such module is found, it is invoked to handle
the authentication.

** System accounting can be turned off by specifying `system no;' in
`acct' section of raddb/config. 

** New configuration statement 'load-module' allows to load arbitrary
Scheme modules.

** Detailed log filenames are configurable via `detail-file-name'
statements in `auth' and `acct' sections of raddb/config.

** Support for Guile versions prior to 1.6 has been withdrawn.

* Bugfixes

** Allow to omit port numbers in `listen' statements (raddb/config), as
described in the documentation.
    


Version 1.2:

* Important notes for upgrading from earlier versions of GNU Radius

** Upgrade your dictionaries.

This version does a very strict semantic checking on the dictionary
files. Old dictionaries may produce lots of warnings and
errors, therefore you are advised to upgrade them. To do so,
run `make update-dictionary'. It will update your main 
(raddb/dictionary) and secondary dictionaries (raddb/dict/*),
preserving any changes you did to the former. 

Notice, that, starting from this version, the default main dictionary
file will not contain any attribute declarations, instead it will
merely include the necessary dictionaries. However, you may add your
declarations to it and modify it to your linking.

Any modifications to the dictionary files in raddb/dict are now
strongly discouraged, as these files will be updated by each new
release. Put your modifications to the main dictionary file only.
This file will remain unchanged during subsequent installs.

Two new statements are provided to help you customize your dictionary.
See section `** Dictionary' below.

** Update your raddb/config and raddb/rewrite

The rewrite procedures are no longer stored in a single file
`radiusd/rewrite', instead they may occupy several source file.
The rewrite procedures shipped with this version of GNU Radius
are installed in ${prefix}/share/radius/1.2/rewrite. We strongly
advise you to add the following statement to your raddb/config:

rewrite {
        load "checknas.rw";
        load "log-hook.rw";
};

You may then remove the old raddb/rewrite file, *unless* you have
modified it.

If you have modified it, follow the procedure described below:

1. Load this file, by adding the following statement to the above
`rewrite' block:

	load "rewrite";

2. Make sure to remove from this file or rename the following functions:
check_snmp_u, check_snmp_s, decimal, check_max_finger, check_unix.

** Tailor log file format.

The format of log file is now fully customizable via a special Rewrite
procedures, called "hooks" (search for logging-hook in the text below
for the detailed description). To revert to logfile format used in
versions 1.0-1.1, add the following two statements to your raddb/config,
right after `logging {' line:

       	prefix-hook "compat_log_prefix";
	suffix-hook "compat_log_suffix";

** Using Guile extensions

If you are using Guile extensions, notice that the default location
for Radius scheme modules is now ${prefix}/share/radius/1.2/scheme.
	
* New features:

** Support for Status-Server request.

** Support for Tunnel Protocol (RFC 2865)

** The main configuration file raddb/config allows to specify several
Rewrite source files to be loaded instead of the single predefined
file raddb/rewrite.

** Arbitrary Rewrite expressions may be used in A/V pairs, for example:

DEFAULT	     Auth-Type = SQL
	     Framed-IP-Address =
	           "=10.10.10.1 + %[NAS-Port-Id]/16 + %[NAS-Port-Id]"

** Test shell mode considerably improved.

** The internal attributes do not appear in detailed logs. The
special flag 'l' has been added to the dictionary syntax to
turn this feature off. Thus, for example:

ATTRIBUTE Orig-User-Name  2101	string	- [L--R-R]l

will make Orig-User-Name attribute visible in detailed log files.

** Changes to Rewrite language

*** usual dotted-quad notation may be used to represent IP numbers.

*** a number of new built-in functions is added

*** #pragma regexp allows to alter the type of regular expressions
used. The rest of line following `#pragma regexp' must consist of
whitespace-separated regexp modifiers. Each regexp modifier is one
of the following keywords:

  extended      Use extended regular expressions
  icase         Ignore case
  newline       Match-any-character operators don't match a newline.

optionally preceded by plus, indicating enabling the feature, or

radwho/radwho.c
/* This file is part of GNU Radius.
   Copyright (C) 2000,2001,2002,2003,2004 Free Software Foundation, Inc.

   Written by Sergey Poznyakoff

   GNU Radius is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
  
   GNU Radius is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with GNU Radius; if not, write to the Free Software Foundation, 
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */

#if defined(HAVE_CONFIG_H)
# include <config.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pwd.h>
#include <errno.h>
#include <sys/stat.h>
#include <sys/utsname.h>
#include <ctype.h>
#include <netinet/in.h>

#include <common.h>
#include <radius/radutmp.h>
#include <radius/radargp.h>

void local_who();
void radius_who();
void print_header();
int want_rad_record(struct radutmp *rt);

/* UTMP stuff. Uses utmpx on svr4 */
#if defined(__svr4__) || defined(__sgi)  
#  include <utmpx.h>
#  include <sys/fcntl.h>
#  define utmp utmpx
#  define UT_NAMESIZE   32
#  define UT_LINESIZE   32
#  define UT_HOSTSIZE   257
#  undef UTMP_FILE
#  define UTMP_FILE UTMPX_FILE
#  undef WTMP_FILE
#  define WTMP_FILE WTMPX_FILE
#else
#  include <utmp.h>
#endif
#ifdef __osf__
#  define UT_NAMESIZE   32
#  define UT_LINESIZE   32
#  define UT_HOSTSIZE   64
#endif
#if defined(__FreeBSD__) || defined(__NetBSD__) || defined(__OpenBSD__) || defined(bsdi)
#  ifndef UTMP_FILE
#    define UTMP_FILE "/var/run/utmp"
#  endif
#  define ut_user ut_name
#endif

#define P_CONSOLE -1  /* Special radutmp type value for local users */

int  fingerd;             /* Are we run as fingerd */
int  secure;              /* Secure mode: do not answer queries w/o username */
int  showlocal;           /* Display local users as well */
int  display_header = 1;  /* Display header line */
int  showall;             /* Display all records */

char *username = NULL;

char *filename = NULL;    /* radutmp filename */
char *eol = "\n";         /* line delimiter */

static struct {
	char *name;
	char *fmt;
} fmtdef[] = {
	{ "default",
	  "(login 10 Login) (gecos 17 Name) (protocol 5 Proto) (nas-port 5 TTY) (time 9 When) (nas-address 9 From) (framed-address 16 Location)" },
	{ "sid",
	  "(login 10 Login) (session-id 17 SID) (protocol 5 Proto) (nas-port 5 TTY) (time 9 When) (nas-address 9 From) (framed-address 16 Location)" },
	{ "clid",
	  "(login 10 Login) (clid 17 CLID) (protocol 5 Proto) (nas-port 5 TTY) (time 9 When) (nas-address 9 From) (framed-address 16 Location)" },
	{ "long",
	  "(login 32 Login) (session-id 32 SID) (protocol 5 Proto) (nas-port 5 Port) (time 27 Date) (nas-address 32 NAS) (clid 17 CLID) (duration 7 Duration) (framed-address 16 Location) (realm 16 Realm)" },
	{ "gnu",
	  "User: (login)(newline)\
In real life: (gecos)(newline)\
Logged in: (time :format %c)(newline)\
NAS: (nas-address)(newline)\
Port: (nas-port)(newline)\
CLID: (clid)(newline)\
Protocol: (protocol)(newline)\
Session ID: (session-id)(newline)\
Uptime: (duration)(newline)\
Assigned IP: (framed-address)(newline)\
Realm: (realm)(newline)" },

	{ NULL }
};

static char *
lookup_format(char *name)
{
	int i;
	for (i = 0; fmtdef[i].name; i++)
		if (strcmp(fmtdef[i].name, name) == 0)
			return fmtdef[i].fmt;
	return name;
}

char *fmtspec = NULL;
format_data_t *form;

const char *argp_program_version = "radwho (" PACKAGE ") " VERSION;
static char doc[] = N_("display who is logged on by Radius");

static struct argp_option options[] = {
        {NULL, 0, NULL, 0,
         N_("radwho specific switches:"), 0},
        {"all", 'A', NULL, 0,
         N_("print all entries, not only active ones"), 0},
        {"calling-id", 'c', NULL, 0,
         N_("display CLID in second column"), 0},
        {"date-format", 'D', N_("DATEFMT"), 0,
         N_("change date representation format"), 0},
        {"empty", 'e', N_("STRING"), 0,
         N_("print STRING instead of an empty column"), 0},
        {"file", 'f', N_("FILE"), 0,
         N_("Use FILE instead of /var/log/radutmp"), 0},
        {"finger", 'F', NULL, 0,
         N_("act as a finger daemon"), 0},
        {"no-header", 'H', NULL, 0,
         N_("do not display header line"), 0},
        {"session-id", 'i', NULL, 0,
         N_("display session ID in the second column"), 0},
        {"ip-strip-domain", 'I', NULL, 0,
         N_("display hostnames without domain part"), 0},
        {"long", 'l', NULL, 0,
         N_("Long output. All fields will be printed."),
         0},
        {"local-also", 'u', NULL, 0,
         N_("display also local users"), 0},
        {"no-resolve", 'n', NULL, 0,
         N_("do not resolve hostnames."), 0},
        {"format", 'o', N_("FORMAT"), 0,
         N_("change output format"), 0},
        {"secure", 's', NULL, 0,
         N_("secure mode: requires that the username be specified"), 0},
        {NULL, 0, NULL, 0, NULL, 0}
};

static error_t
parse_opt(int key, char *arg, struct argp_state *state)
{
        switch (key) {
        case 'A': /* display all entries */
                showall++;
                break;
        case 'c': /* CLID instead of GECOS */
                fmtspec = lookup_format("clid");
                break;
        case 'D': /* Date format */
                printutmp_date_format = optarg;
                break;
        case 'e': /* empty field replacement */
                printutmp_empty_string = optarg;
                break;
        case 'f': /* filename */
                filename = optarg;
                break;
        case 'F':
                fingerd++;
                break;
        case 'H': /* Disable header line */
                display_header = 0;
                break;
        case 'i': /* Display SID instead of GECOS */
                fmtspec = lookup_format("sid");
                break;
        case 'I': /* Ipaddr format */
                /*FIXME set_ip_format(optarg);*/
                break;
        case 'l': /* long output */
                fmtspec = lookup_format("long");
                break;
        case 'n':
                resolve_hostnames = 0;
                break;
        case 'o':
                fmtspec = lookup_format(optarg);
                break;
        case 's':
                secure++;
                break;
        case 'u':
                showlocal++;
                break;
        default:
                return ARGP_ERR_UNKNOWN;
        }
        return 0;
}

static struct argp argp = {
        options,
        parse_opt,
        NULL,
        doc,
        rad_common_argp_child,
        NULL, NULL
};


int
main(int  argc, char **argv)
{
        char inbuf[128];
        char *path;
        char *p, *q;
        int index;

        grad_app_setup();
        if (grad_argp_parse(&argp, &argc, &argv, 0, &index, NULL))
                return 1;

        if (!fmtspec)
                fmtspec = getenv("RADWHO_FORMAT");

        if (!fmtspec)
                fmtspec = lookup_format("default");

	form = grad_utent_compile_form(fmtspec);
	if (!form)
		exit(1);

        if (!filename)
                filename = radutmp_path;
        
        /* Read the dictionary files */
        grad_dict_init();
        /* Read the "naslist" file. */
        path = grad_mkfilename(radius_dir, RADIUS_NASLIST);
	if (grad_nas_read_file(path))
                exit(1);
        grad_free(path);
        /* Read realms */
        path = grad_mkfilename(radius_dir, RADIUS_REALMS);
	grad_read_realms(path, 0, 0, NULL);
        grad_free(path);
        
        /*
         *      See if we are "fingerd".
         */
        if (strstr(argv[0], "fingerd")) 
                fingerd++;

        if (fingerd) {
                eol = "\r\n";
                /*
                 *      Read first line of the input.
                 */
                fgets(inbuf, sizeof(inbuf), stdin);
                p = inbuf;
                while(*p == ' ' || *p == '\t') p++;
                if (*p == '/' && *(p + 1)) p += 2;
                while(*p == ' ' || *p == '\t') p++;
                for(q = p; *q && *q != '\r' && *q != '\n'; q++)
                        ;
                *q = 0;
                if (*p)
                        username = p;

                /*
                 *      See if we fingered a specific user.
                 */
                if (secure && username == 0) {
                        printf(_("must provide username\n"));
                        exit(1);
                }
        }

        if (showlocal)
                local_who();

        radius_who();

        fflush(stdout);
        fflush(stderr);

        return 0;
}

void
tty_to_port(struct radutmp *rt, char *tty)
{
        char *p;

        p = tty + strlen(tty) - 1; 
        while (p >= tty && isdigit(*p))
                p--;
        rt->nas_port = atoi(p+1);
        rt->porttype = 0;/*FIXME*/
}

void
local_who()
{
        FILE  *fp;
        struct utmp ut;
        struct radutmp rt;
        
        if ((fp = fopen(UTMP_FILE, "r")) == NULL) {
                grad_log(L_ERR, _("can't open file: %s"),
                         UTMP_FILE);
                return;
        }

        print_header();

        memset(&rt, 0, sizeof(rt));
        rt.nas_address = rt.framed_address = htonl(INADDR_LOOPBACK);
        
        while(fread(&ut, sizeof(ut), 1, fp) == 1) {
#ifdef USER_PROCESS
                if (ut.ut_user[0] && ut.ut_line[0] &&
                    ut.ut_type == USER_PROCESS) {
#else
                if (ut.ut_user[0] && ut.ut_line[0]) {
#endif
                        rt.type = P_CONSOLE;
                        strncpy(rt.login, ut.ut_name, RUT_NAMESIZE);
                        strncpy(rt.orig_login, ut.ut_host, RUT_NAMESIZE);
#if defined __svr4__ || defined __sgi
                        rt.time = ut.ut_xtime;
#else
                        rt.time = ut.ut_time;
#endif
                        tty_to_port(&rt, ut.ut_line);
                        if (want_rad_record(&rt)) 
                                grad_utent_print(form, &rt, 1);
                }
        }
        fclose(fp);
}

void
radius_who()
{
        radut_file_t file;
        struct radutmp *up;
	
        print_header();

        /*
         *      Show the users logged in on the terminal server(s).
         */
        if ((file = rut_setent(filename, 0)) == NULL)
                return ;

        while (up = rut_getent(file)) {
		if (up->type == P_ACCT_DISABLED) 
			printf(_("System accounting is disabled\n"));
		else if (want_rad_record(up)) 
                        grad_utent_print(form, up, 1);
        }
        rut_endent(file);
}

void
print_header()
{
        if (display_header) {
                grad_utent_print_header(form);
                display_header = 0;
        }
}

int
want_rad_record(struct radutmp *rt)
{
        if ((username && strcmp(rt->login, username))
	    || rt->type == P_ACCT_ENABLED) 
                return 0;
        
        switch (showall) {
        case 0:
                return rt->type != P_IDLE;
        case 1:
                return rt->login[0] != 0;
        case 2:
        default:
                return (rt->type == P_IDLE && rt->login[0] != 0);
        }
}


